File: 16.3-sum-closest.py
#
# @lc app=leetcode id=16 lang=python3
#
# [16] 3Sum Closest
#

# @lc code=start
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()

        res = nums[0] + nums[1] + nums[2]

        for i in range (len(nums) - 2):
            if nums[i] == nums[i - 1]:
                continue

            left, right = i + 1, len(nums) - 1

            while left < right:
                curr_sum = nums[i] + nums[left] + nums[right]

                if curr_sum == target:
                    return curr_sum
                elif abs(curr_sum - target) < abs(res - target):
                    res = curr_sum
                
                if curr_sum < target:
                    left += 1
                else:
                    right -= 1
            
        return res


        
        

        
# @lc code=end

File: 16.3-sum-closest.java
/*
 * @lc app=leetcode id=16 lang=java
 *
 * [16] 3Sum Closest
 */

// @lc code=start

import java.util.Arrays;

class Solution {
    public int threeSumClosest(int[] nums, int target) {
        
        Arrays.sort(nums);

        int numsSize = nums.length;

        int closetSum = nums[0] + nums[1] + nums[2];

        for (int i = 0; i < numsSize - 2; i++) {

            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            int left = i + 1;
            int right = numsSize - 1;

            while (left < right) {

                int currentSum = nums[i] + nums[left] + nums[right];

                if (Math.abs(currentSum - target) < Math.abs(closetSum - target)) {
                    closetSum = currentSum;
                }

                if (currentSum == target) {
                    return currentSum;
                } else if (currentSum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return closetSum;
    }
}
// @lc code=end

